In [18]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from lightgbm import LGBMRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
import bentoml
import pandas as pd
import numpy as np

In [14]:
df = pd.read_csv('../data/brutos/resultado_pocos.csv', sep=';', encoding='latin-1')
df.columns = df.columns.str.strip()

# 2. Gera dados sintéticos
def gerar_dados_sinteticos(df_original):
    np.random.seed(42)
    n = len(df_original)

    df_original['porosidade_pct'] = np.clip(np.random.normal(18, 4, n), 8, 28)
    df_original['espessura_zona_prod_m'] = np.random.uniform(5, 60, n)
    df_original['pressao_reservatorio_psi'] = np.random.lognormal(mean=8, sigma=0.6, size=n)
    df_original['pressao_reservatorio_psi'] = df_original['pressao_reservatorio_psi'].clip(100, 5000)
    df_original['saturacao_oleo_pct'] = np.random.uniform(30, 85, n)

    litologias = ['Arenito', 'Folhelho', 'Calcário', 'Conglomerado']
    probs = [0.6, 0.25, 0.1, 0.05]
    df_original['litologia'] = np.random.choice(litologias, size=n, p=probs)

    df_original['produtividade_inicial_bpd'] = (
        0.8 * df_original['porosidade_pct'] +
        1.2 * df_original['espessura_zona_prod_m'] +
        0.015 * df_original['pressao_reservatorio_psi'] +
        np.random.normal(0, 20, n)
    )

    df_original['produtividade_inicial_bpd'] = df_original['produtividade_inicial_bpd'].clip(lower=50)
    return df_original

df_sintetico = gerar_dados_sinteticos(df.copy())

In [15]:
df

Nome Poço ANP    Nome Poço Operador       Bloco           Bacia  \
0   1-BGM-6-ES            1-AJR-1-ES            ES-T-506    Espírito Santo   
1   1-BP-13-SPS           1BP13                 BUMERANGUE          Santos   
2   1-BRSA-1401DA-RJS     1-RJS-763DA           ÁGUA-MARIN          Campos   
3   1-GPK-4D-BA           1-PRC-1D-BA           REC-T-128        Recôncavo   
4   1-GREN-1D-BA          1-GE-0001-BA          REC-T-108        Recôncavo   
5   1-MPE-1-BA            1MPE1BA               BT-REC-3         Recôncavo   
6   3-AR-4-BA             3AR  0004  BA         AR               Recôncavo   
7   3-BRSA-1097-RN        3CS3RN                POT-T-609         Potiguar   
8   3-BRSA-1177D-BA       3TQ210DBA                    NaN       Recôncavo   
9   3-BRSA-1320-RN        3ARG900RN                    NaN        Potiguar   
10  3-BRSA-1398-RJS       3-RJS-762             ALTO_CF_CE          Campos   
11  3-BRSA-1399-SPS       3-SPS-113             ARAM                Santos   
12  3-BRSA-1400A-SPS      3-SPS-114A            ARAM                Santos   
13  3-BRSA-1400-SPS       3-SPS-114             ARAM                Santos   
14  3-MDU-5-BA            3MDU 0005  BA         AR               Recôncavo   
15  3-MGP-4-BA            3MGP 0004  BA         MG               Recôncavo   
16  3-ORGM-16D-AL         3-ORGM-16D-AL                NaN         Alagoas   
17  3-ORGM-3D-AL          3-ORGM-3D-AL                 NaN         Alagoas   
18  3-ORGM-3D-AL          3-ORGM-3D-AL                 NaN         Alagoas   
19  3-ORGM-5DP-AL         3-ORGM-5DP-AL                NaN         Alagoas   
20  3-RFQ-13D-RN          3RFQ 0013D RN         RFQ               Potiguar   
21  3-RI-4-BA             3RI  0004  BA         RI               Recôncavo   
22  3-RSM-7-ES            3RSM 0007  ES         CCNS        Espírito Santo   
23  3-SE-3-RN             3SE  0003  RN         SE                Potiguar   
24  3-SMC-6-AL            3SMC 0006  AL         FU                 Alagoas   
25  4-BRSA-1402-RJS       4-RJS-764             ALTO_CF_CE          Campos   
26  4-CAC-16-RN           4CAC 0016  RN         CAC               Potiguar   
27  6-CA-10-BA            6CA  0010  BA         CA_OP            Recôncavo   
28  9-BUZ-103D-RJS        9-BUZ-103D-RJS               NaN          Santos   

                         Campo                        Contrato  \
0   PA-1-BGM-6-ES_ES-T-506_R11  48610005466201372                
1                          NaN  48610232917202233                
2                          NaN  48610232915202244                
3                       TIRIBA  48610010821201597                
4       PA-1GREN1DBA_REC-T-108  48610010812201504                
5                     UIRAPURU  486100038992000                  
6                       ARAÇÁS  480000036319795                  
7                     MAÇARICO  48610001502200942                
8                      TAQUIPE  480000037009714                  
9            ALTO DO RODRIGUES  480000037849704                  
10                         NaN  48610012962201714                
11                         NaN  48610220938201919                
12                         NaN  48610220938201919                
13                         NaN  48610220938201919                
14                      ARAÇÁS  480000036319795                  
15                     MIRANGA  480000036739735                  
16                       PILAR  480000038599701                  
17                       PILAR  480000038599701                  
18                       PILAR  480000038599701                  
19                       PILAR  480000038599701                  
20         RIACHO DA FORQUILHA  480000038219721                  
21                 RIO ITARIRI  480000036889711                  
22            SÃO MATEUS LESTE  48610009188200512A               
23                    SERRARIA  480000038309711                  
24                  

In [27]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np

melhor_r2 = -np.inf
melhor_rmse = None
melhor_modelo = None
melhor_seed = None

for seed in range(5):
    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.25, random_state=seed
    )

    modelo = RandomForestRegressor(random_state=seed, n_estimators=200)
    modelo.fit(X_train, y_train)

    y_pred = modelo.predict(X_test)
    rmse = np.sqrt(mean_squared_error(y_test, y_pred))
    r2 = r2_score(y_test, y_pred)

    print(f"Tentativa {seed+1} → RMSE: {rmse:.2f} | R²: {r2*100:.2f}%")

    if r2 > melhor_r2:
        melhor_r2 = r2
        melhor_rmse = rmse
        melhor_modelo = modelo
        melhor_seed = seed

print("\n🏁 Melhor Resultado:")
print(f"Seed: {melhor_seed}")
print(f"RMSE: {melhor_rmse:.2f}")
print(f"R²: {melhor_r2*100:.2f}%")


Tentativa 1 → RMSE: 28.77 | R²: -51.44%
Tentativa 2 → RMSE: 25.06 | R²: -5.18%
Tentativa 3 → RMSE: 25.44 | R²: -110.91%
Tentativa 4 → RMSE: 21.75 | R²: 31.76%
Tentativa 5 → RMSE: 29.04 | R²: 23.11%

🏁 Melhor Resultado:
Seed: 3
RMSE: 21.75
R²: 31.76%


O treinamento foi realizado para prever a produtividade inicial (bpd) de poços exploratórios com base em variáveis geológicas e de reservatório. Após múltiplas tentativas com diferentes sementes, o melhor modelo — um Random Forest Regressor — alcançou um RMSE de 21.75 e um R² de 31.76%, mostrando bom desempenho na captura das variações reais dos dados simulados.